lib/wallaroo/core/source/connector_source/connector_framed_source_notify.pony

```diff
@@ -82,91 +254,348 @@ class ConnectorSourceNotify[In: Any val]
+        source.close()
+        return _continue_perhaps(source)
+
+      | let m: cwm.NotifyMsg =>
+        ifdef "trace" then
+          @printf[I32]("TRACE: got NotifyMsg: %lu %s %lu\n".cstring(),
+            m.stream_id, m.stream_name.cstring(), m.point_of_ref)
+        end
+        if _fsm_state isnt _ProtoFsmStreaming then
+          return _to_error_state(source, "Bad protocol FSM state")
+        end
 
-          let msg_uid = _msg_id_gen()
+        if _pending_notify.contains(m.stream_id) or
+           _active_streams.contains(m.stream_id) or
+           _pending_close.contains(m.stream_id)
+        then
+          // This notifier is already handling this stream
+          // So reject directly
+          send_notify_ack(false, m.stream_id, m.point_of_ref)
+        else
+          _process_notify(where source=source, stream_id=m.stream_id,
+            stream_name=m.stream_name, point_of_reference=m.point_of_ref)
+        end
 
-          // TODO: We need a way to determine the key based on the policy
-          // for any particular connector. For example, the Kafka connector
-          // needs a way to provide the Kafka key here.
-          let initial_key = msg_uid.string()
+      | let m: cwm.NotifyAckMsg =>
+        ifdef "trace" then
+          @printf[I32]("TRACE: got NotifyAckMsg\n".cstring())
+        end
+        return _to_error_state(source, "Invalid message: notify_ack")
 
-          // TOOD: We need a way to assign watermarks based on the policy
-          // for any particular connector.
-          if ingest_ts > _watermark_ts then
-            _watermark_ts = ingest_ts
-          end
+      | let m: cwm.MessageMsg =>
+        // check that we're in state that allows processing messages
+        if _fsm_state isnt _ProtoFsmStreaming then
+          return _to_error_state(source, "Bad protocol FSM state")
+        end
+        if not cwm.FlagsAllowed(m.flags) then
+          return _to_error_state(source, "Bad MessageMsg flags")
+        end
 
-          if decoded isnt None then
-            _runner.run[In](_pipeline_name, pipeline_time_spent, decoded,
-              consume initial_key, ingest_ts, _watermark_ts, _source_id,
-              source, _router, msg_uid, None, decode_end_ts,
-              latest_metrics_id, ingest_ts, _metrics_reporter)
-          else
-            (true, ingest_ts)
-          end
+        // try to process message
+        if not _active_streams.contains(m.stream_id) then
+          return _to_error_state(source, "Bad stream_id " + m.stream_id.string())
         else
-          @printf[I32](("Unable to decode message at " + _pipeline_name +
-            " source\n").cstring())
-          ifdef debug then
-            Fail()
+          try
+            let s = _active_streams(m.stream_id)?
+            let msg_id = try
+              m.message_id as cwm.MessageId
+            else
+              0
+            end
+
+            if (msg_id > 0) and (msg_id <= s.last_seen) then
```

Comment:
What happens if twice in a row `m.message_id as cwm.MessageId` causes an error and the msg_id is set to 0?